src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+            Assert.check(this.refs == null);
+            Assert.check(this.depth == -1);
+            Assert.check(this.callStack.isEmpty());
+            Assert.check(this.pendingWarning == null);
+            Assert.check(this.invocations.isEmpty());
+        }
+        return true;
+    }
+
+// Ref's
+
+    /** Represents a location that could possibly hold a 'this' reference.
+     *
+     *  <p>
+     *  If not "direct", the reference is found through at least one indirection.
+     */
+    private abstract static class Ref {
+
+        private final int depth;
+        private final boolean direct;
+
+        Ref(int depth, boolean direct) {
+            this.depth = depth;
+            this.direct = direct;
+        }
+
+        public int getDepth() {
+            return this.depth;
+        }
+
+        public boolean isDirect() {
+            return this.direct;
+        }
+
+        @Override
+        public int hashCode() {
+            return this.getClass().hashCode()
+                ^ Integer.hashCode(this.depth)
+                ^ Boolean.hashCode(this.direct);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this)
+                return true;
+            if (obj == null || obj.getClass() != this.getClass())
+                return false;
+            final Ref that = (Ref)obj;
+            return this.depth == that.depth
+              && this.direct == that.direct;
+        }
+
+        @Override
+        public String toString() {
+            final ArrayList<String> properties = new ArrayList<>();
+            this.addProperties(properties);
+            return this.getClass().getSimpleName()
+              + "[" + properties.stream().collect(Collectors.joining(",")) + "]";
+        }
+
+        protected void addProperties(ArrayList<String> properties) {
+            properties.add("depth=" + this.depth);
+            properties.add(this.direct ? "direct" : "indirect");
+        }
+    }
+
+    /** A reference from the current 'this' instance.
+     */
+    private static class ThisRef extends Ref {
+
+        ThisRef(boolean direct) {
+            super(0, direct);
+        }
+
+        public static ThisRef direct() {
+            return new ThisRef(true);
+        }
+
+        public static ThisRef indirect() {
+            return new ThisRef(false);
+        }
+    }
+
+    /** A reference from the current outer 'this' instance.
+     */
+    private static class OuterRef extends Ref {
+
+        OuterRef(boolean direct) {
+            super(0, direct);
+        }
+
+        public static OuterRef direct() {
+            return new OuterRef(true);
+        }
+
+        public static OuterRef indirect() {
+            return new OuterRef(false);
+        }
+    }
+
+    /** A reference from the expression that was just evaluated.
+     *  In other words, a reference that's sitting on top of the stack.
+     */
+    private static class ExprRef extends Ref {
```

Comment:
Maybe I'm wrong - but it seems that the only thing we need to track is whether the top of stack (e.g. last evaluated expression) has a direct reference or indirect reference (or none). Do we really need a set for this? (this seems the same as for ThisRef which is used inside method, which can be one of the same three options). While I do see value for tracking which variables are aliases for this (either direct or indirect), it seems like (at least on a superficial look) that expression refs might be replaced by a visitor field/parameter.

Reply:
> Do we really need a set for this?

There are surely other ways to model things. But I got myself really confused trying to build more complicated models.

What I ended up with is this simple model that works:
* There is a set of `Ref` subclasses that model the various types of 'this' references possible: `OuterRef`, `ExprRef`, etc.
* There is a singleton `this.refs`, which just is a `Set<Ref>`, containing the 'this' references that are alive at the current moment
* As we "execute" code, all we need to do is update the `this.refs` based on what the current bit of code does

E.g....

When a variable is assigned a value that has a reference, we add a `VarRef` for that variable.

When we leave a scope, we remove any `Ref`'s that are no longer in scope.

Before executing a method, we add `Ref`'s for the method receiver and parameter.

When we return from a non-void method, we convert any `ReturnRef` into a `ExprRef`.

Etc.

THAT I can understand.

I don't see converting the above into a bitmap or whatever as worth the additional complexity. We're not programming in perl here.